### PR TITLE
[Metricbeat] Update Couchdb to v2

### LINKED ISSUE #16352: metricbeat couchdb module not working with couch 2.x
metricbeat version: 7.5.2
couchdb version: 2.1.0
system: Linux, CentOS-7.4
discussion: https://discuss.elastic.co/t/metricbeat-couchdb-module-not-working/219094/3

The metricbeat couchdb module does not work for couchdb-2.x. This version has a different _stats response than 1.x, and the module is zeroing out all numeric values in its elastic docs.

On a related note, to get it working at all, you need to set the module's "hosts" entries to something like "localhost:5984/_node/_local/_stats". This is undocumented and confusing, as "hosts" does not suggest that you need to provide the stats endpoint there.

### PR DESCRIPTION
**Pull request by sayden**
## What does this PR do?

Updates module to use a Strategy patern and decide between v1 and v2 Couchdb parsing. To do so, it has to do a pre HTTP request to root path `/` to know which version is deployed on each fetch (done in fetch to allow hot upgrades in container environment).

I have removed `testdata` tests because one of them was a bit unnecessary with our current `testdata` implementation (a timeout test which was testing `testdata` _framework_ more than the module itself). The other `testdata` doesn't work anymore because more than one request is necessary now to check version.

Because we can only test in CI with a single version, I maintained current Dockerfile

## Why is it important?

ATM, we were only parsing v1 metrics and v3 is going to be released soon. This PR introduces v2 parsing maintaining compatibility with our previous version.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
~~- [ ] I have made corresponding change to the default configuration files~~
- [ ] I have added tests that prove my fix is effective or that my feature works

## How to test this PR locally

* Launch v1 and v2 containers
```
docker run -d --name couchdb16 -p 15984:5984 couchdb:1.6
docker run -d --name couchdb23 -p 25984:5984 -p 25986:5986 couchdb:2.3
```
* Setup Metricbeat Couchdb module to point to both servers `./metricbeat modules enable couchdb` and leave config similar to this:
```
- module: couchdb
  metricsets: ["server"]
  period: 10s
  hosts: ["localhost:15984", "localhost:25984/_node/_local/_stats", "localhost:25984/_node/nonode@nohost/_stats", "localhost:25986"]
```
* Check that events from v1 and v2 are being received in, for example, Kibana. You can see version in field `ecs.service.version` and the `ecs.service.id` must be the same for all paths of v2.
* Extra mile, check dashboards, they must maintain compatibility and show info of both servers.

## Related issues

- Closes elastic/beats#16352
- Relates elastic/beats#10354

## Screenshots
I checked that dashboard still works.

image

### CODE REVIEW

File: metricbeat/module/couchdb/server/server.go [hunk 1]
```diff
@@ -61,23 +71,108 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
...
+	}
+
+	v, err := getVersionFromInfo(info)
+	if err != nil {
+		return errors.Wrap(err, "error trying to parse version info")
+	}
+
+	switch v.major {
+	case "1":
+		m.fetcher = &V1{}
```
Comment (mtojek) [inline metricbeat/module/couchdb/server/server.go]: Do you think this can be optimized to not create a new instance every time the `Fetch` is called?

File: metricbeat/module/couchdb/server/server.go [hunk 2]
```diff
@@ -61,23 +71,108 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
...
+		http:          http,
+		fetcher:       nil,
 	}, nil
 }
 
 // Fetch methods implements the data gathering and data conversion to the right
 // format. It publishes the event which is then forwarded to the output. In case
 // of an error set the Error field of mb.Event or simply call report.Error().
 func (m *MetricSet) Fetch(reporter mb.ReporterV2) error {
+	info, err := m.getInfoFromCouchdbHost(m.Host())
```
[the remark this answers is not in the document]
Comment (sayden) [reply metricbeat/module/couchdb/server/server.go]: Historically this was the preferred way to achieve this. Even in very heavy operations like opening connections to SQL databases. Frankly, I never liked very much the idea but I understood the implications (a long living SQL connection that may drop and leave the beat requiring a restart.

Revalidation using schema errors or any other more complex stuff is against the common "guides" (written nowhere) in Beats of trying to have explicit, easy to reason / easy to code / easy to maintain codebase. In Couchdb it's not a very good idea because the differences are very very small and can lead to unexpected results. An example are incomplete JSON returns by design or a mix of errors from the http transport layer with the database layer that can make your life a hell :smile: 

TLDR: I'll disagree and commit and move it to `New` :wink:

File: metricbeat/module/couchdb/server/server.go (hunk 1 above)
Comment (sayden) [reply metricbeat/module/couchdb/server/server.go]: See above

File: metricbeat/module/couchdb/server/server.go [hunk 3]
```diff
@@ -61,23 +71,108 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
...
+
+type version struct {
+	major string
+	minor string
+	patch string
+}
+
+// Parse db version from HTTP response
+func getVersionFromInfo(info *CommonInfo) (*version, error) {
+	versionSlice := strings.Split(info.Version, ".")
```
[the remark this answers is not in the document]
Comment (sayden) [reply metricbeat/module/couchdb/server/server.go]: Good catch!

File: metricbeat/module/couchdb/server/server.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mtojek) [reply metricbeat/module/couchdb/server/server.go]: Hmm... I wonder if it can lead to a potential problem - beats app starts before couchdb so it can't determine the API version.

**Comment by sayden** [comment]
jenkins, test this

File: metricbeat/module/couchdb/server/server.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ycombinator) [reply metricbeat/module/couchdb/server/server.go]: > Hmm... I wonder if it can lead to a potential problem - beats app starts before couchdb so it can't determine the API version.

Yes! We ran into this exact issue with a couple of the stack monitoring modules in Metricbeat and had to move all API calls into `Fetch`. See https://github.com/elastic/beats/issues/15258 (fixed by https://github.com/elastic/beats/pull/15270) and https://github.com/elastic/beats/issues/15276 (fixed by https://github.com/elastic/beats/pull/15306).

One solution I've seen in the past is caching such "infrequently changing" information for some period of time to save on checking it every `Fetch()` period. But that can introduce its own issues such as staleness so it might not be worth it. In general I err towards keeping it simple until it starts to be a problem.

[the remark this answers is not in the document]
Comment (mtojek) [reply metricbeat/module/couchdb/server/server.go]: Regarding the offline discussion - I think you can move the code determining API version just before calling the right fetcher. Both fetchers should be stateless, so you freely initialize them in `New()` and get rid of any unnecessary mem allocations in `Fetch`.

**Comment by sayden** [comment]
I have done a last update by fetching on `New` and retrying on `Fetch` in case something of service unavailable on `New`.

Let's park the implementation here until we have _any_ feedback from community and users before over engineering this more. The real problem here is that a JSON parsing from v1 vs v2 won't raise any error. A custom JSON parser must be implemented to prevent silent errors on hot upgrades. As this is a very short-living corner case, it's reasonable to stop here.

File: metricbeat/module/couchdb/server/server.go [hunk 4]
```diff
@@ -76,8 +95,76 @@ func (m *MetricSet) Fetch(reporter mb.ReporterV2) error {
 		return errors.Wrap(err, "error in http fetch")
 	}
 
-	event, _ := eventMapping(content)
-	reporter.Event(mb.Event{MetricSetFields: event})
+	if m.fetcher == nil {
```
Comment (mtojek) [inline metricbeat/module/couchdb/server/server.go]: nit: I would hide this condition inside retrieveFetcher. The provider creates an instance if it's not ready.

Comment (sayden) [reply metricbeat/module/couchdb/server/server.go]: Moved it inside, thanks!
